### ISSUE TITLE
Sporadic crash when during on my TV/Receiver due to double close() on the same fd in ActiveAE

### ISSUE THREAD

**Issue by yasij** [xbmc/xbmc#16258]
### Describe the bug
Here is a clear and concise description of what the problem is:

Kodi crashes sporadically when I turn on my TV and receiver since I upgraded to Ubuntu 19.04. There is a double close somewhere on the ActiveAE thread which is racing against the PeripBusUSBUdev. libudev has an assert when close returns EBADF. When I turn the TV on, it triggers udev_enumerate_scan_devices. This has been difficult to debug due to the sporadic nature of the race condition. I've straced kodi-x11 with close() calls, but if I add openat to the list, it disturbs the timing enough that I haven't been able to reproduce with that strace attached. It happens at least 50% of the time turning my TV on with no strace attached.

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

Kodi should not crash sporadically when turning on my TV.

## Actual Behavior

Kodi crashes sporadically when turning on my TV.

### To Reproduce

1. Boot machine with auto login to kodi. I'm using ALSA HDMI audio. The EDID is fixed.
2. Turn my TV on. This triggers the receiver on via CEC.
3. Kodi crashes.

### Debuglog

The debuglog can be found here:
(Debugging is not on for this log because it disturbs the timing enough to mess with the race condition, and makes it hard to reproduce):
[kodi_crashlog-20190609_171212.log](https://github.com/xbmc/xbmc/files/3274163/kodi_crashlog-20190609_171212.log)

Stack strace from the core dump, it asserts inside libudev due to the double close on the other thread:
[kodistacktrace3.log](https://github.com/xbmc/xbmc/files/3274165/kodistacktrace3.log)
List of thread names for the tids:
[kodithreadlist3.txt](https://github.com/xbmc/xbmc/files/3274167/kodithreadlist3.txt)
Tail of the an strace showing the double close on the ActiveAE thread:
[kodistracetail.log](https://github.com/xbmc/xbmc/files/3274174/kodistracetail.log)

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [x] Linux
 - [ ] OSX
 - [ ] Raspberry-Pi
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: Ubuntu 19.04
 - Kodi version: 18.2

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by yasij** [comment]
I did an strace with open, openat, close, creat and ioctl to try to better see what's going on. For the ActiveAE thread: The openat is on /dev/snd/controlC0. There is a double close of that file descriptor later on the thread. There must be a bad call to snd_ctl_close somewhere. I'll try to get a debug build of Kodi to trace those calls.

**Comment by yasij** [comment]
I finally had time to debug this more. It's a bug in jack2. I'm not even using jack on my system. However, with alsa-plugins >= 1.1.7, the ALSA to JACK plugin layer is enabled by default. Ubuntu 19.04 ships with alsa-plugins 1.1.8. Ubuntu 18.10 had alsa-plugins 1.1.6 which explains the behavior change. I'll send a patch to jack2.

**Comment by fritsch** [comment]
Could you summarize the issue with jack in more detail here? Perhaps there is something in kodi we can improve.

**Comment by yasij** [comment]
Fix merged in jack: https://github.com/jackaudio/jack2/commit/dad4b5702782eef3bd66e3c3f4fefaaae3571208

Here are the stack traces for the two closes in libjack, and the addr2line decoded line numbers:
[jackclose1.txt](https://github.com/xbmc/xbmc/files/3306518/jackclose1.txt)
[jackclose2.txt](https://github.com/xbmc/xbmc/files/3306519/jackclose2.txt)
[jackdecoded1.txt](https://github.com/xbmc/xbmc/files/3306520/jackdecoded1.txt)
[jackdecoded2.txt](https://github.com/xbmc/xbmc/files/3306521/jackdecoded2.txt)

Kodi is calling EnumerateDevice over the list of ALSA sinks, and then TryDevice is opening each of them. For the jack plugin that Ubuntu 19.04 enables by default, that tries to open jack. However, libjack has a double close when it fails to connect to jackd because it closes the socket on error, and then closes it again in the cleanup code. I'm not running jackd. I don't even have it installed. That 2nd close races against the udev enumerator in kodi to cause the sporadic crash.

I think the only real workaround from the kodi side would be to ignore the ALSA jack plugin unless a user explicitly wants it.

**Comment by fritsch** [comment]
Thanks much. I think an Ubuntu Bugreport should be preferred then.

**Comment by martin3000** [comment]
In the flatpak version of kodi 21.2 I get this error:
`Assertion 'close_nointr(fd) != -EBADF' failed at src/basic/fd-util.c:75, function safe_close(). Aborting.`
Can you help me to narrow down the error? How can I find the module name and how do I start strace to find the file descriptor which is closed twice?

**Comment by neo1973** [comment]
@martin3000: Please look here: https://github.com/flathub/tv.kodi.Kodi/issues/320
